File: Duality.R
library(shiny)
library(plotly)

## ----- Moments (swap in your own) -----
mu <- c(1.06, 1.04, 1.08, 1.03)  # gross means
n  <- length(mu)
set.seed(42)
A <- matrix(rnorm(n*n), n, n)
Sigma <- A %*% t(A)
Sigma <- Sigma / (max(diag(Sigma)) * 150)
one <- rep(1, n)
SigInv <- solve(Sigma)

## ----- Helpers -----
mv_frontier_sd <- function(R){
  A <- as.numeric(t(one) %*% SigInv %*% one)
  B <- as.numeric(t(one) %*% SigInv %*% mu)
  C <- as.numeric(t(mu)  %*% SigInv %*% mu)
  D <- A*C - B^2
  sqrt((A*R^2 - 2*B*R + C)/D)
}
tangency_point <- function(Rf){
  z <- SigInv %*% (mu - Rf*one)
  if(all(abs(z) < 1e-12)){
    w <- (SigInv %*% one) / as.numeric(t(one) %*% SigInv %*% one)   # GMV fallback
  } else {
    w <- z / as.numeric(t(one) %*% z)
  }
  mean <- as.numeric(t(w) %*% mu)
  sd   <- sqrt(as.numeric(t(w) %*% Sigma %*% w))
  c(sd = sd, mean = mean)
}
hj_sigma_min <- function(mbar){
  v <- one - mbar*mu
  sqrt(as.numeric(t(v) %*% SigInv %*% v))
}

## ----- Static grids -----
# GMV
gmv_w  <- (SigInv %*% one) / as.numeric(t(one) %*% SigInv %*% one)
gmv_er <- as.numeric(t(gmv_w) %*% mu)

# Build a dense R-grid, compute (sd, er)
R_grid <- seq(0.95*min(min(mu), gmv_er),
              1.05*max(max(mu), gmv_er), length.out = 1200)
MV_all <- data.frame(sd = mv_frontier_sd(R_grid), er = R_grid)

# Split by ER relative to GMV, order each half by sd increasing
MV_left  <- subset(MV_all, er <= gmv_er)
MV_right <- subset(MV_all, er >= gmv_er)
MV_left  <- MV_left[order(MV_left$sd), ]
MV_right <- MV_right[order(MV_right$sd), ]

# Stitch into ONE continuous path: left→GMV→right (drop duplicate GMV row)
MV_curve <- rbind(MV_left, MV_right[-1, , drop = FALSE])

# HJ frontier grid
HJ_axis <- seq(0.6, 1.4, length.out = 400)
HJ <- data.frame(mbar = HJ_axis, sd = sapply(HJ_axis, hj_sigma_min))

## ----- UI -----
ui <- fluidPage(
  titlePanel("MV–HJ Duality"),
  sidebarLayout(
    sidebarPanel(
      sliderInput("mbar", "E[m] (moves both panels; Rf = 1/E[m])",
                  min = 0.70, max = 1.30, value = 1.00, step = 0.005, width = "100%")
    ),
    mainPanel(
      plotlyOutput("mvPlot", height = "380px"),
      plotlyOutput("hjPlot", height = "360px")
    )
  )
)

## ----- Server -----
server <- function(input, output, session){
  
  output$mvPlot <- renderPlotly({
    Rf <- 1.0 / input$mbar
    tp <- tangency_point(Rf)
    
    plot_ly() |>
      add_lines(data = MV_curve, x = ~sd, y = ~er, name = "MV Frontier") |>
      add_markers(x = tp["sd"], y = tp["mean"], name = "Tangency", marker = list(size = 9)) |>
      add_markers(x = mv_frontier_sd(gmv_er), y = gmv_er, name = "GMV",
                  marker = list(size = 8, symbol = "diamond")) |>
      layout(
        title = paste0("Top: Rf≈", round(Rf,3),
                       "  →  Tangency (sd≈", round(tp["sd"],3),
                       ", E[R]≈", round(tp["mean"],3), ")"),
        xaxis = list(title = "Portfolio SD"),
        yaxis = list(title = "Expected Gross Return")
      )
  })
  
  output$hjPlot <- renderPlotly({
    mbar <- input$mbar
    sdm  <- hj_sigma_min(mbar)
    
    plot_ly() |>
      add_lines(data = HJ, x = ~mbar, y = ~sd, name = "HJ Frontier") |>
      add_markers(x = mbar, y = sdm, name = "Current E[m]", marker = list(size = 9)) |>
      layout(
        title = paste0("Bottom: min σ(m) at E[m] = ", round(mbar,3),
                       "  ≈  ", round(sdm,3)),
        xaxis = list(title = "E[m]"),
        yaxis = list(title = "σ(m)")
      )
  })
}

shinyApp(ui, server)

File: Duality.py
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# ---------- Moments (you can replace with your estimates) ----------
mu = np.array([1.06, 1.04, 1.08, 1.03])   # gross means
n = len(mu)
rng = np.random.default_rng(42)
A = rng.normal(size=(n, n))
Sigma = A @ A.T
Sigma = Sigma / (np.max(np.diag(Sigma)) * 150.0)
one = np.ones(n)
SigInv = np.linalg.inv(Sigma)

# ---------- MV frontier helpers ----------
def mv_frontier(R):
    A = one @ SigInv @ one
    B = one @ SigInv @ mu
    C = mu @ SigInv @ mu
    D = A*C - B**2
    var = (A*R**2 - 2*B*R + C) / D
    return np.sqrt(var)

def tangency_point(Rf):
    z = SigInv @ (mu - Rf*one)
    if np.allclose(z, 0):
        w = (SigInv@one)/(one@SigInv@one)
    else:
        w = z/(one@z)
    mean = w @ mu
    sd   = np.sqrt(w @ Sigma @ w)
    return sd, mean

# ---------- HJ frontier ----------
def hj_sigma_min(mbar):
    v = (one - mbar*mu)
    return np.sqrt(v @ SigInv @ v)

# Grids
R_grid = np.linspace(0.95*min(mu.min(), ((SigInv@one)/(one@SigInv@one))@mu),
                     1.05*max(mu.max(), ((SigInv@one)/(one@SigInv@one))@mu), 200)
mv_sd = mv_frontier(R_grid)

mbar_grid = np.linspace(0.70, 1.30, 51)     # slider values (E[m])
Rf_grid   = 1.0/mbar_grid
hj_sd_grid = np.array([hj_sigma_min(x) for x in np.linspace(0.6,1.4,200)])
hj_mbar_axis = np.linspace(0.6,1.4,200)

# ---------- Build figure with two panels ----------
fig = make_subplots(rows=2, cols=1, shared_xaxes=False,
                    subplot_titles=("Mean–Variance Frontier (Top)", "HJ Frontier (Bottom)"))

# static lines
fig.add_trace(go.Scatter(x=mv_sd, y=R_grid, mode='lines', name="MV Frontier"), row=1, col=1)
fig.add_trace(go.Scatter(x=hj_mbar_axis, y=hj_sd_grid, mode='lines', name="HJ Frontier"), row=2, col=1)

# moving markers (initialized)
sd0, er0 = tangency_point(Rf_grid[0])
m0, s0   = mbar_grid[0], hj_sigma_min(mbar_grid[0])

tan_trace = go.Scatter(x=[sd0], y=[er0], mode='markers', name="Tangency", marker=dict(size=9))
hj_trace  = go.Scatter(x=[m0],  y=[s0], mode='markers', name="Current E[m]", marker=dict(size=9))

fig.add_trace(tan_trace, row=1, col=1)
fig.add_trace(hj_trace,  row=2, col=1)

fig.update_xaxes(title_text="Portfolio SD", row=1, col=1)
fig.update_yaxes(title_text="Expected Gross Return", row=1, col=1)
fig.update_xaxes(title_text="E[m]", row=2, col=1)
fig.update_yaxes(title_text="σ(m)", row=2, col=1)

# ---------- Slider frames (update markers as mbar changes) ----------
steps = []
for i, mbar in enumerate(mbar_grid):
    Rf = Rf_grid[i]
    sd_t, er_t = tangency_point(Rf)
    sd_m = hj_sigma_min(mbar)
    step = dict(
        method="update",
        args=[{"data": [
            # 0: MV line, 1: HJ line, 2: Tangency point, 3: HJ point
            fig.data[0],               # unchanged MV line
            fig.data[1],               # unchanged HJ line
            go.Scatter(x=[sd_t], y=[er_t], mode='markers', name="Tangency", marker=dict(size=9)),
            go.Scatter(x=[mbar], y=[sd_m], mode='markers', name="Current E[m]", marker=dict(size=9)),
        ]},
        {"annotations": []}],
        label=f"E[m]={mbar:.3f}\nRf≈{Rf:.3f}"
    )
    steps.append(step)

fig.update_layout(
    sliders=[dict(active=0, steps=steps, x=0.05, xanchor="left", y= -0.08, len=0.9)],
    height=800,
    showlegend=True,
    title="MV–HJ Duality (use slider to change E[m]; Rf = 1/E[m])"
)

# Save a single interactive HTML file
fig.write_html("mv_hj_interactive.html", include_plotlyjs="cdn", full_html=True)
print("Wrote mv_hj_interactive.html")
